lib/gems/pending/util/miq_file_storage.rb

```diff
@@ -0,0 +1,177 @@
+    #
+    # In the block form of the method, only the remote_uri is required, and it
+    # is assumed the input will be a generated in the executed block (most
+    # likely an external process) to a unix pipe that can be written to.  The
+    # pipe generated by this method and passed in to the block as a file
+    # location to the `input_stream`).
+    #
+    # In the non-block form, a source must be provioded as the first argument
+    # either as an IO object that can be read from, or a file path, and the
+    # second argument is the remote_uri as in the block form.
+    #
+    # An additional argument in both forms as the last argument is `byte_count`
+    # can also be included.  If passed, it will be assumed that the resulting
+    # input will be split, and the naming for the splits will be:
+    #
+    #   - filename.00001
+    #   - filename.00002
+    #   ...
+    #
+    # Block form:
+    #
+    #   nfs_session.add("path/to/file", "200M") do |input_stream|
+    #     `pg_dump -f #{input_stream} vmdb_production`
+    #   end
+    #
+    # Non-block form:
+    #
+    #   nfs_session.add("path/to/local_file", "path/to/remote_file")
+    #   nfs_session.add("path/to/local_file", "path/to/remote_file", "200M")
+    #
+    def add(*upload_args, &block)
+      initialize_upload_vars(*upload_args)
+      handle_io_block(&block)
+      mkdir(File.dirname(@remote_file_path))
+      if byte_count
+        upload_splits
+      else
+        upload_single(@remote_file_path)
+      end
+    ensure
+      reset_vars
+    end
+    alias upload add
+
+    def mkdir(dir) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{self.class}##{__callee__} is not defined"
+    end
+
+    def download(local_file, remote_file_uri) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{self.class}#download is not defined"
+    end
+
+    private
+
+    def upload_single(file) # rubocop:disable Lint/UnusedMethodArgument
+      raise NotImplementedError, "#{self.class}#upload_single is not defined"
+    end
+
+    def upload_splits
+      @position = 0
+      until source_input.eof?
+        upload_single(next_split_filename)
+        @position += byte_count
+      end
+    end
+
+    def initialize_upload_vars(*upload_args)
+      upload_args.pop if (@byte_count = parse_byte_value(upload_args.last))
+      @remote_file_path = upload_args.pop
+
+      if upload_args.empty?
+        require "pty"
+
+        puts
+        puts "setting @source_input and @input_writer via PTY..."
+        @source_input, @input_writer = PTY.open
+        puts "set to: "
+        puts "  @source_input: #{@source_input.inspect}"
+        puts "  @input_writer: #{@input_writer.inspect}"
+      else
+        source        = upload_args.pop
+        @source_input = source.kind_of?(IO) ? source : File.open(source, "r")
+      end
+    end
+
+    def parse_byte_value(bytes)
+      match = bytes.to_s.match(BYTE_HASH_MATCH) || return
```

Comment:
This seems a bit fragile for my tastes. I would rather go with a straight options hash or kwargs rather than determining the meaning of a parameter based on a home-grown regex.

Reply:
Oh, sorry, should I have implemented the fun case logic that is included as part of `xstrtol.c`:

https://github.com/coreutils/gnulib/blob/5ff4c9ba8b2002538380e3930dfc6b0bd5f4b297/lib/xstrtol.c

Which is what is used by `coreutils` in GNU `split` to parse the `--bytes` flag?

...

:trollface: 

(I may or may not have went on a lengthy hunt to find how this was done in between travis runs just for the sake of this troll...)

* * *

Okay, but for the real ( 📖 ) answer, I assume parsing `"200M"` instead of `209715200` is not what your beef is with, but the implementation of the "`:call_sig:`" of `#add`, correct?

So, I am not married to this implementation, but the one reason I choose it was to avoid having to change every instance where `MiqGenericMountSession` (and friends) is used to conform to the new interface.  I do agree that `kwargs` probably is a much better choice, but trying to maintain the current interface for now while adding the file splitting functionality in as well, and this was the best I could come up with.

A middle ground that would avoid the `arg` shifting in `#initialize_upload_vars` would be to just allow `local_io` to be passed in as `nil`.  The reason I am not a huge fan of that is I am usually not a fan of the first arg being optionally `nil`, because then you lead to calls that look like this:

```ruby
MiqLocalMountSession.add nil, "path/to/destination/file", "200M" do |output_file|
  `pg_dump -f #{output_file}...`
end
```
And I think that just muddies the interface, and makes the caller wonder what is going on with the `nil` for the argument.

But again, not married to this implementation, just my personal preference for a ideal-ish implementation and rational.

Reply:
I guess to play a bit of devil's advocate for myself, I could probably implement the `kwargs` interface for the top level class method that I want to implement anyway, and just keep `#add` conforming to the current form arg structure.  🤔

Reply:
Being that this comment is a month old, I am going to update and say that I didn't go through with refactoring this.  I have spent a crap ton of time validating that this does indeed work (in unit tests and [the test suite gist](https://gist.github.com/NickLaMuro/8438015363d90a2d2456be650a2b9bc6)), and since it is "crunch time", I am going to argue we try and address this later.

Reply:
Sounds good to me, I'll resolve this.